ledger/participant-integration-api/src/main/scala/platform/index/InMemoryUserManagementService.scala

```diff
@@ -0,0 +1,99 @@
+import com.daml.ledger.participant.state.index.v2.UserManagementService
+import com.daml.ledger.participant.state.index.v2.UserManagementService._
+
+import scala.concurrent.Future
+import scala.collection.mutable
+
+class InMemoryUserManagementService extends UserManagementService {
+  import InMemoryUserManagementService._
+
+  override def createUser(user: User, rights: Set[UserRight]): Future[Result[Unit]] = Future.successful {
+    putIfAbsent(UserInfo(user, rights)) match {
+      case Some(_) => Left(UserExists(user.id))
+      case None => Right(())
+    }
+  }
+
+  override def getUser(id: String): Future[Result[User]] = Future.successful {
+    lookup(id) match {
+      case Some(userInfo) => Right(userInfo.user)
+      case None => Left(UserNotFound(id))
+    }
+  }
+
+  override def deleteUser(id: String): Future[Result[Unit]] = Future.successful {
+    dropExisting(id) match {
+      case Some(_) => Right(())
+      case None => Left(UserNotFound(id))
+    }
+  }
+
+  override def grantRights(id: String, granted: Set[UserRight]): Future[Result[Set[UserRight]]] = Future.successful {
+    lookup(id) match {
+      case Some(userInfo) =>
+        val newlyGranted = granted.diff(userInfo.rights) // faster than filter
+        // we're not doing concurrent updates -- assert as backstop and a reminder to handle the collision case in the future
+        assert(replaceInfo(userInfo, userInfo.copy(rights = userInfo.rights ++ newlyGranted)))
+        Right(newlyGranted)
+      case None =>
+        Left(UserNotFound(id))
+    }
+  }
+
+  override def revokeRights(id: String, revoked: Set[UserRight]): Future[Result[Set[UserRight]]] = Future.successful {
+    lookup(id) match {
+      case Some(userInfo) =>
+        val effectivelyRevoked = revoked.intersect(userInfo.rights) // faster than filter
+        // we're not doing concurrent updates -- assert as backstop and a reminder to handle the collision case in the future
+        assert(replaceInfo(userInfo, userInfo.copy(rights = userInfo.rights -- effectivelyRevoked)))
+        Right(effectivelyRevoked)
+      case None =>
+        Left(UserNotFound(id))
+    }
+  }
+
+  override def listUserRights(id: String): Future[Result[Set[UserRight]]] = Future.successful {
+    lookup(id) match {
+      case Some(userInfo) => Right(userInfo.rights)
+      case None => Left(UserNotFound(id))
+    }
+  }
+
+  // Underlying mutable map to keep track of UserInfo state.
+  // Structured so we can use a ConcurrentHashMap (to more closely mimic a real implementation, where performance is key).
+  // We synchronize on a private object (the mutable map), not the service (which could cause deadlocks).
+  // (No need to mark state as volatile -- rely on synchronized to establish the JMM's happens-before relation.)
+  private val state: mutable.Map[String, UserInfo] = mutable.Map(AdminUser.toStateEntry)
+  private def lookup(id: String) = state.synchronized { state.get(id) }
```

Comment:
@adriaanm-da : if you use a `ConcurrentHashMap` why do you need the `synchronized`? That means all reads and writes get sequentially ordered doesn't it?

Reply:
Yes, if we switched to a concurrent map, the `synchronized` usages should be removed (and the `val` storing the concurrent map should be final so that it's published properly). I meant the logic was structured so that switching to a lock-free implementation like CHM would be straightforward (using operations like `putIfAbsent` and `replace`). I don't know if the additional complexity is warranted, though -- I'd start with a simple solution like this, using `synchronize` on a mutable map, and see if it scales well enough.